crates/uv-publish/src/lib.rs

```diff
@@ -0,0 +1,784 @@
+            struct ErrorBody {
+                code: String,
+            }
+
+            if let Ok(structured) = serde_json::from_str::<ErrorBody>(&body) {
+                structured.code
+            } else {
+                body
+            }
+        } else {
+            body
+        }
+    }
+}
+
+pub fn files_for_publishing(
+    paths: Vec<String>,
+) -> Result<Vec<(PathBuf, DistFilename)>, PublishError> {
+    let mut seen = HashSet::new();
+    let mut files = Vec::new();
+    for path in paths {
+        for entry in glob(&path).map_err(|err| PublishError::Pattern(path, err))? {
+            let entry = entry?;
+            if !seen.insert(entry.clone()) {
+                continue;
+            }
+            if !entry.is_file() {
+                continue;
+            }
+            let Some(filename) = entry.file_name().and_then(|filename| filename.to_str()) else {
+                continue;
+            };
+            let filename = DistFilename::try_from_normalized_filename(filename)
+                .ok_or_else(|| PublishError::InvalidFilename(entry.clone()))?;
+            files.push((entry, filename));
+        }
+    }
+    // TODO(konsti): Should we sort those files, e.g. wheels before sdists because they are more
+    // certain to have reliable metadata, even though the metadata in the upload API is unreliable
+    // in general?
+    Ok(files)
+}
+
+/// Upload a file to a registry.
+///
+/// Returns `true` if the file was newly uploaded and `false` if it already existed.
+pub async fn upload(
+    file: &Path,
+    filename: &DistFilename,
+    registry: &Url,
+    client: &BaseClient,
+    username: Option<&str>,
+    password: Option<&str>,
+) -> Result<bool, PublishError> {
+    let form_metadata = form_metadata(file, filename)
+        .await
+        .map_err(|err| PublishError::PublishPrepare(file.to_path_buf(), err))?;
+    let request = build_request(
+        file,
+        filename,
+        registry,
+        client,
+        username,
+        password,
+        form_metadata,
+    )
+    .await
+    .map_err(|err| PublishError::PublishPrepare(file.to_path_buf(), err))?;
+
+    let response = request.send().await.map_err(|err| {
+        PublishError::PublishSend(file.to_path_buf(), registry.clone(), err.into())
+    })?;
+
+    handle_response(registry, response)
+        .await
+        .map_err(|err| PublishError::PublishSend(file.to_path_buf(), registry.clone(), err))
+}
+
+/// Calculate the SHA256 of a file.
+fn hash_file(path: impl AsRef<Path>) -> Result<String, io::Error> {
+    // Ideally, this would be async, but in case we actually want to make parallel uploads we should
+    // use `spawn_blocking` since sha256 is cpu intensive.
+    let mut file = BufReader::new(File::open(path.as_ref())?);
+    let mut hasher = Sha256::new();
+    io::copy(&mut file, &mut hasher)?;
+    Ok(format!("{:x}", hasher.finalize()))
+}
+
+// Not in `uv-metadata` because we only support tar files here.
```

Comment:
It seems like this could still make sense in `uv-metadata`. Why not exatly?

Reply:
In uv-metadata, we support all the files we get from external sources, that is all kinds of tar compressions and notably `.zip` files, which are historically popular as source distributions. Here, we specifically only support `.tar.gz`, the now only officially allowed extension. If you were to reuse this function in any place outside uploading, it would do the wrong thing and reject existing source distributions.